[JSC] Add linked-on-after quirk for canDeclareGlobalFunction()

shvaikalesh opened the pull request:
#### 355e9d25900a300375b1b68576f926278d8b5903
<pre>
[JSC] Add linked-on-after quirk for canDeclareGlobalFunction()
<a href="https://bugs.webkit.org/show_bug.cgi?id=267199">https://bugs.webkit.org/show_bug.cgi?id=267199</a>
&lt;<a href="https://rdar.apple.com/120965147">rdar://120965147</a>&gt;

Reviewed by Mark Lam.

This change adds linked-on-after quirk for behavior introduced in <a href="https://webkit.org/b/260487">https://webkit.org/b/260487</a>,
preventing TypeError from being thrown if canDeclareGlobalFunction() fails, yet only in global code.
Instead of throwing, we bring back old non-compliant behavior of removing non-configurable global property.

* Source/JavaScriptCore/jsc.cpp:
Just like WTR / DRT are calling enableAllSDKAlignedBehaviors(), make JSC shell call it as well.
It&apos;s necessary for linkedOnOrAfterSDKWithBehavior() to work locally and on EWS.

* Source/JavaScriptCore/runtime/ProgramExecutable.cpp:
(JSC::requiresCanDeclareGlobalFunctionQuirk):
(JSC::ProgramExecutable::initializeGlobalProperties):
* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.cpp:
(WTF::computeSDKAlignedBehaviors):
* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:

Canonical link: <a href="https://commits.webkit.org/273773@main">https://commits.webkit.org/273773@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7713b2212fa416680676f5298141ed79fa86444f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/36641 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/15581 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/38869 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/39300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/18056 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/12660 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/39300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/37202 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/18056 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/38869 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/39300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/18056 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/38869 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/40546 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/30756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/18056 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/38869 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/40546 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/36415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/11773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/12660 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/40546 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/38869 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/43200 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12184 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/43200 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4738 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/12659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -808,6 +809,10 @@ class GlobalObject final : public JSGlobalObject {
         if (Options::useDollarVM()) {
+#if PLATFORM(COCOA)
+            enableAllSDKAlignedBehaviors();
+#endif
```
Why is `enableAllSDKAlignedBehaviors()` only needed for the jsc shell binary in testing mode?  Why not all the time?

shvaikalesh replied:
Fixed!

MenloDorian reviewed (approved):
r=me

Commit:
[JSC] Add linked-on-after quirk for canDeclareGlobalFunction()
https://bugs.webkit.org/show_bug.cgi?id=267199
<rdar://120965147>

Reviewed by Mark Lam.

This change adds linked-on-after quirk for behavior introduced in https://webkit.org/b/260487,
preventing TypeError from being thrown if canDeclareGlobalFunction() fails, yet only in global code.
Instead of throwing, we bring back old non-compliant behavior of removing non-configurable global property.

* Source/JavaScriptCore/jsc.cpp:
Just like WTR / DRT are calling enableAllSDKAlignedBehaviors(), make JSC shell call it as well.
It's necessary for linkedOnOrAfterSDKWithBehavior() to work locally and on EWS.

* Source/JavaScriptCore/runtime/ProgramExecutable.cpp:
(JSC::requiresCanDeclareGlobalFunctionQuirk):
(JSC::ProgramExecutable::initializeGlobalProperties):
* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.cpp:
(WTF::computeSDKAlignedBehaviors):
* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:

Canonical link: https://commits.webkit.org/273773@main
